bubblepicker/src/main/java/com/igalata/bubblepicker/physics/Engine.kt

```diff
@@ -18,67 +21,77 @@ object Engine {
         set(value) {
             field = value
             bubbleRadius = interpolate(0.1f, 0.25f, value / 100f)
-            gravity = interpolate(20f, 80f, value / 100f)
-            standardIncreasedGravity = interpolate(500f, 800f, value / 100f)
         }
-    var centerImmediately = false
-    private var standardIncreasedGravity = interpolate(500f, 800f, 0.5f)
-    private var bubbleRadius = 0.17f
+    private var bubbleRadius = 0.15f
 
     private val world = World(Vec2(0f, 0f), false)
-    private val step = 0.0005f
+    private val step = 1.0f / 60.0f
     private val bodies: ArrayList<CircleBody> = ArrayList()
     private var borders: ArrayList<Border> = ArrayList()
     private val resizeStep = 0.005f
     private var scaleX = 0f
     private var scaleY = 0f
     private var touch = false
-    private var gravity = 6f
-    private var increasedGravity = 55f
-    private var gravityCenter = Vec2(0f, 0f)
-    private val currentGravity: Float
-        get() = if (touch) increasedGravity else gravity
     private val toBeResized = ArrayList<Item>()
-    private val startX
-        get() = if (centerImmediately) 0.5f else 2.2f
-    private var stepsCount = 0
+
+    override fun endContact(p0: Contact?) {
+    }
+
+    override fun beginContact(p0: Contact) {
+        borders.forEach { border ->
+            if (border.itemBody == p0.fixtureA.body || border.itemBody == p0.fixtureB.body) {
+                val circle = if (border.itemBody == p0.fixtureA) p0.fixtureA else p0.fixtureB
+                var vel = circle.body.linearVelocity
+                vel = if (border.view == Border.HORIZONTAL) {
+                    Vec2(vel.x, -vel.y)
+                } else {
+                    Vec2(-vel.x, vel.y)
+                }
+                circle.body.linearVelocity = vel
+                return@forEach
+            }
+        }
+    }
+
+    override fun preSolve(p0: Contact?, p1: Manifold?) {
+    }
+
+    override fun postSolve(p0: Contact?, p1: ContactImpulse?) {
+    }
 
     fun build(bodiesCount: Int, scaleX: Float, scaleY: Float): List<CircleBody> {
         val density = interpolate(0.8f, 0.2f, radius / 100f)
         for (i in 0..bodiesCount - 1) {
-            val x = if (Random().nextBoolean()) -startX else startX
-            val y = if (Random().nextBoolean()) -0.5f / scaleY else 0.5f / scaleY
-            bodies.add(CircleBody(world, Vec2(x, y), bubbleRadius * scaleX, (bubbleRadius * scaleX) * 1.3f, density))
+            val x = Random().nextFloat() - 0.5f
+            val y = Random().nextFloat() - 0.5f
+            val vx = (if (Random().nextBoolean()) -0.01f else 0.01f) * Random().nextFloat() / scaleY
+            val vy = (if (Random().nextBoolean()) -0.01f else 0.01f) * Random().nextFloat() / scaleY
+            bodies.add(CircleBody(world, Vec2(x, y), bubbleRadius * scaleX, (bubbleRadius * scaleX) * 1.1f, density, Vec2(vx, vy)))
```

Comment:
Can you explain what the change in these lines does?

Reply:
`x` and `y` are the initial positions for the circles.
`vx` and `vy` are the initial force we're going to apply to the bubbles to get them moving.